src/wallet/deterministic_wallet.go

```diff
@@ -0,0 +1,337 @@
+func (w *DeterministicWallet) GetEntryAt(i int) Entry {
+	return w.Entries[i]
+}
+
+// GetEntry returns entry of given address
+func (w *DeterministicWallet) GetEntry(a cipher.Address) (Entry, bool) {
+	return w.Entries.get(a)
+}
+
+// HasEntry returns true if the wallet has an Entry with a given cipher.Address.
+func (w *DeterministicWallet) HasEntry(a cipher.Address) bool {
+	return w.Entries.has(a)
+}
+
+// GenerateAddresses generates addresses
+func (w *DeterministicWallet) GenerateAddresses(num uint64) ([]cipher.Addresser, error) {
+	if num == 0 {
+		return nil, nil
+	}
+
+	if w.Meta.IsEncrypted() {
+		return nil, ErrWalletEncrypted
+	}
+
+	var seckeys []cipher.SecKey
+	var seed []byte
+	if len(w.Entries) == 0 {
+		seed, seckeys = cipher.MustGenerateDeterministicKeyPairsSeed([]byte(w.Meta.Seed()), int(num))
+	} else {
+		sd, err := hex.DecodeString(w.Meta.LastSeed())
+		if err != nil {
+			return nil, fmt.Errorf("decode hex seed failed: %v", err)
+		}
+		seed, seckeys = cipher.MustGenerateDeterministicKeyPairsSeed(sd, int(num))
+	}
+
+	w.Meta.setLastSeed(hex.EncodeToString(seed))
+
+	addrs := make([]cipher.Addresser, len(seckeys))
+	makeAddress := w.Meta.AddressConstructor()
+	for i, s := range seckeys {
+		p := cipher.MustPubKeyFromSecKey(s)
+		a := makeAddress(p)
+		addrs[i] = a
+		w.Entries = append(w.Entries, Entry{
+			Address: a,
+			Secret:  s,
+			Public:  p,
+		})
+	}
+	return addrs, nil
+}
+
+// GenerateSkycoinAddresses generates Skycoin addresses. If the wallet's coin type is not Skycoin, returns an error
+func (w *DeterministicWallet) GenerateSkycoinAddresses(num uint64) ([]cipher.Address, error) {
+	if w.Meta.Coin() != CoinTypeSkycoin {
+		return nil, errors.New("GenerateSkycoinAddresses called for non-skycoin wallet")
+	}
+
+	addrs, err := w.GenerateAddresses(num)
+	if err != nil {
+		return nil, err
+	}
+
+	skyAddrs := make([]cipher.Address, len(addrs))
+	for i, a := range addrs {
+		skyAddrs[i] = a.(cipher.Address)
+	}
+
+	return skyAddrs, nil
+}
+
+// reset resets the wallet entries and move the lastSeed to origin
+func (w *DeterministicWallet) reset() {
+	w.Entries = Entries{}
+	w.Meta.setLastSeed(w.Meta.Seed())
+}
+
+// ScanAddresses scans ahead N addresses, truncating up to the highest address with a non-zero balance.
+// If any address has a nonzero balance, it rescans N more addresses from that point, until a entire
+// sequence of N addresses has no balance.
+func (w *DeterministicWallet) ScanAddresses(scanN uint64, tf TransactionsFinder) error {
+	if w.Meta.IsEncrypted() {
+		return ErrWalletEncrypted
+	}
+
+	if scanN == 0 {
+		return nil
+	}
+
+	w2 := w.Clone().(*DeterministicWallet)
+
+	nExistingAddrs := uint64(len(w2.Entries))
+	nAddAddrs := uint64(0)
+	n := scanN
+	extraScan := uint64(0)
+
+	for {
+		// Generate the addresses to scan
+		addrs, err := w2.GenerateSkycoinAddresses(n)
+		if err != nil {
+			return err
+		}
+
+		// Find if these addresses had any activity
+		active, err := tf.AddressesActivity(addrs)
```

Comment:
Maybe we can have a method in the `tf` interface to check addresses activity from the end to first. And returns the index of the first encountered address with balance. We actually don't care if the addresses have balance before the last active address. In this way, the method don't have to go through all addresses unless all are empty. And because we already know the address index, we don't have to go through the addresses again in the following code to get the `keepNum`.

Reply:
Can make that a separate issue (optimize scanner)

Reply:
https://github.com/skycoin/skycoin/issues/2437